From c5691747eaed7c13e0f7dcbffa411efa94ed42e5 Mon Sep 17 00:00:00 2001
From: Sora Su <baxiche@gmail.com>
Date: Thu, 23 Oct 2025 14:34:12 +0800
Subject: [PATCH] oem-meta: Add call to the fw-versions-sd-retimer service

The fw-versions-sd-retimer service is triggered by the POST_COMPLETED
and POST_ENDED events. This is a downstream-only implementation, as
the service is not supported upstream.

Change-Id: Iffcf7ac3473689785fcd44b38081f52533dac552
Signed-off-by: Sora Su <baxiche@gmail.com>
Signed-off-by: Eric Yang <eric.yang.wiwynn@gmail.com>
Upstream-Status: Inappropriate
---
 .../file_io_type_event_log.cpp                | 26 +++++++++++++++++++
 1 file changed, 26 insertions(+)

diff --git a/oem/meta/libpldmresponder/file_io_type_event_log.cpp b/oem/meta/libpldmresponder/file_io_type_event_log.cpp
index 5029f19..a9ad75e 100644
--- a/oem/meta/libpldmresponder/file_io_type_event_log.cpp
+++ b/oem/meta/libpldmresponder/file_io_type_event_log.cpp
@@ -962,6 +962,25 @@ static void reportError(const message&, const std::string& event,
 {
     pldm::utils::reportError(event.c_str());
 }
+
+static void updateRetimerFwVersion(const message&, const std::string&,
+                                   const std::string& slotNum)
+{
+    static constexpr auto SYSTEMD_SERVICE = "org.freedesktop.systemd1";
+    static constexpr auto SYSTEMD_ROOT = "/org/freedesktop/systemd1";
+    static constexpr auto SYSTEMD_INTERFACE =
+        "org.freedesktop.systemd1.Manager";
+
+    /* The retimer is not available temporary after fw update.
+       We should update fw version on d-bus when post-complete. */
+    auto bus = sdbusplus::bus::new_default();
+    auto method = bus.new_method_call(SYSTEMD_SERVICE, SYSTEMD_ROOT,
+                                      SYSTEMD_INTERFACE, "StartUnit");
+    auto service = std::string("fw-versions-sd-retimer@") + slotNum +
+                   ".service";
+    method.append(service, "replace");
+    bus.call_noreply(method);
+}
 } // namespace report
 
 namespace format
@@ -1125,6 +1144,13 @@ static void addSystemEventLogAndJournal(const message& eventData,
     {
         lg2::error("Unexpected error: {ERROR}", "ERROR", event);
     }
+
+    // Trigger retimer fw version update on POST_ENDED and POST_COMPLETED
+    if (eventType == EventType::POST_ENDED ||
+        eventType == EventType::POST_COMPLETED)
+    {
+        report::updateRetimerFwVersion(eventData, event, slotNum);
+    }
 }
 
 int EventLogHandler::write(const message& eventData)
-- 
2.53.0-Meta

